### PR TITLE
prow: lgtm plugin, remove label on PR synchronize

### PR DESCRIPTION
**Pull request by nlandolfi**
Incorporate the functionality of the lgtm_after_commit munger into the prow lgtm plugin.

Closes https://github.com/kubernetes/test-infra/issues/3795 <-- see that issue for more context as well

### CODE REVIEW

File: prow/plugins/lgtm/lgtm.go [hunk 1]
```diff
@@ -204,3 +207,26 @@ func handle(gc githubClient, log *logrus.Entry, e *event) error {
...
+	case "synchronize":
+		// Don't bother checking if it has the label...it's a race, and we'll have
+		// to handle failure due to not being labeled anyway.
+		if err := gc.RemoveLabel(
+			pe.PullRequest.Base.Repo.Owner.Login,
+			pe.PullRequest.Base.Repo.Name,
+			pe.PullRequest.Number,
+			lgtmLabel,
+		); err != nil {
+			// TODO(nlandolfi): inspect this error?
```
Comment (nlandolfi) [inline prow/plugins/lgtm/lgtm.go]: especially curious for thoughts on what to do here?

File: prow/github/fakegithub/fakegithub.go [hunk 2]
```diff
@@ -195,3 +195,7 @@ func (f *FakeClient) GetFile(org, repo, file, commit string) ([]byte, error) {
 
 	return nil, fmt.Errorf("could not find file %s with ref %s", file, commit)
 }
+
+func LabelID(org, repo string, number int, label string) string {
```
Comment (nlandolfi) [inline prow/github/fakegithub/fakegithub.go]: pulled this out of the functions and exported for use in external libs...right call?

File: prow/plugins/lgtm/lgtm.go [hunk 3]
```diff
@@ -204,3 +207,26 @@ func handle(gc githubClient, log *logrus.Entry, e *event) error {
 	}
 	return nil
 }
+
+func handlePullRequest(gc githubClient, pe github.PullRequestEvent) error {
+	if pe.PullRequest.Merged {
+		return nil
+	}
+
+	switch pe.Action {
```
Comment (spxtr) [inline prow/plugins/lgtm/lgtm.go]: Unindent with
```go
if pe.Action != "synchronize" {
	return nil
}
```

File: prow/plugins/lgtm/lgtm.go (hunk 1 above)
Comment (spxtr) [inline prow/plugins/lgtm/lgtm.go]: What does the API return if you ask it to remove a label from a PR that doesn't have that label? Try it out on a test PR.

**Comment by spxtr** [comment]
The munger also adds a comment saying something along the lines of "removing LGTM after the PR changed," and then it also deletes old versions of that comment when they're no longer relevant. I'm not a huge fan of that comment, but it's something worth considering.

File: prow/plugins/lgtm/lgtm.go (hunk 3 above)
Comment (nlandolfi) [inline prow/plugins/lgtm/lgtm.go]: done!

File: prow/plugins/lgtm/lgtm.go (hunk 1 above)
Comment (nlandolfi) [inline prow/plugins/lgtm/lgtm.go]: returns a 404 Label does not exist, but not sure how well exposed this is when the RemoveLabel returns. I could inspect the error text specifically, but the raw error interface doesn't give me much info...looks like the error code will be something like:

`status code 404 not one of [200 204], body: Label not found`

alternative is to add 404 to list of "ok" return codes in RemoveLabel...

File: prow/github/fakegithub/fakegithub.go (hunk 2 above)
Comment (spxtr) [inline prow/github/fakegithub/fakegithub.go]: It's fine either way. I wish we didn't have fakegithub at all...

File: prow/plugins/lgtm/lgtm.go (hunk 1 above)
Comment (spxtr) [inline prow/plugins/lgtm/lgtm.go]: The way to inspect errors in go is with a type assertion. Check out how it's done with the [assign](https://github.com/kubernetes/test-infra/blob/b4330c6aef47a5f9d99d881439a08333774c297a/prow/github/client.go#L519-L554) method, and how it's used in the [assign plugin](https://github.com/kubernetes/test-infra/blob/b4330c6aef47a5f9d99d881439a08333774c297a/prow/plugins/assign/assign.go#L169-L182). The only tricky thing is that prow's github client retrys 404s a couple times, so we'll need some special casing or something to make that not happen.

File: prow/github/fakegithub/fakegithub.go (hunk 2 above)
Comment (nlandolfi) [inline prow/github/fakegithub/fakegithub.go]: i noticed that github/ has a FakeClient method but it seemed to be added before the fakegithub/ dir so I figured current practice was to use the fakegithub client.

I can avoid using fakegithub by creating my own mock in the test.

Comment (spxtr) [inline prow/github/fakegithub/fakegithub.go]: I prefer fakes per plugin.

**Comment by nlandolfi** [comment]
@spxtr should we add the commenting functionality to the prow plugin?

personal anecdote: on PRs I see "k8s-bot removed label..." notifications in timelines so i think that's sufficient, but also understand wanting to maintain parity between munger and prow..

thoughts?

**Comment by nlandolfi** [comment]
Also I can no longer find our discussion about the RemoveLabel error...but I think the way to go is to modify the RemoveLabels to return a typed error ...

**Comment by spxtr** [comment]
I agree that the timeline entry should be enough.

/ok-to-test

**Comment by nlandolfi** [comment]
let me know if there is anything else on my end!

File: prow/github/client.go [hunk 4]
```diff
@@ -505,15 +505,55 @@ func (c *Client) AddLabel(org, repo string, number int, label string) error {
 	return err
 }
 
+// Use LabelNotFound to indicate that a label is not attached to an issue.
+// For example, removing a label from an issue, when the issue does not have that label.
+type LabelNotFound struct {
```
Comment (kargakis) [inline prow/github/client.go]: I think this was added in your size plugin changes, right?

**Comment by kargakis** [comment]
Please rebase on top of master and bump hook.

File: prow/plugins/lgtm/lgtm.go [hunk 5]
```diff
@@ -204,3 +207,34 @@ func handle(gc githubClient, log *logrus.Entry, e *event) error {
...
+func handlePullRequest(gc ghLabelClient, pe github.PullRequestEvent) error {
+	if pe.PullRequest.Merged {
+		return nil
+	}
+
+	if pe.Action != "synchronize" {
+		return nil
+	}
+
+	// Don't bother checking if it has the label...it's a race, and we'll have
```
Comment (spxtr) [inline prow/plugins/lgtm/lgtm.go]: We should consider checking if it has the label only to save the API token that comes with attempting to remove it.

File: prow/github/client.go (hunk 4 above)
Comment (nlandolfi) [inline prow/github/client.go]: i thought I might have also but I don't see it there. I think I added the file not found error that might be what you are remembering.

File: prow/plugins/lgtm/lgtm.go (hunk 5 above)
Comment (nlandolfi) [inline prow/plugins/lgtm/lgtm.go]: done! PTAL

File: prow/plugins/lgtm/lgtm.go [hunk 6]
```diff
@@ -204,3 +207,56 @@ func handle(gc githubClient, log *logrus.Entry, e *event) error {
...
+
+	var (
+		org    = pe.PullRequest.Base.Repo.Owner.Login
+		repo   = pe.PullRequest.Base.Repo.Name
+		number = pe.PullRequest.Number
+	)
+
+	// Check if it has the label before removing to save the API token
+	// that comes with attempting to remove a label.
+	labels, err := gc.GetIssueLabels(org, repo, number)
```
Comment (spxtr) [inline prow/plugins/lgtm/lgtm.go]: Argh, I misunderstood something. I was thinking you could say `if pe.PullRequest.HasLabel("lgtm") { ... }` but the labels are not sent in the `PullRequestEvent`, so no matter what, you have to spend the token. You can undo this change, you were right before that just trying to remove and ignoring the LabelNotFound error is correct.

Comment (nlandolfi) [inline prow/plugins/lgtm/lgtm.go]: ahhhh gotcha -- will do

File: prow/plugins/lgtm/lgtm.go (hunk 5 above)
Comment (nlandolfi) [inline prow/plugins/lgtm/lgtm.go]: reverted as per more recent discussion

**Comment by nlandolfi** [comment]
ok I reverted to the original where we just remove the label

File: prow/github/client.go (hunk 4 above)
Comment (kargakis) [inline prow/github/client.go]: ah, yeah

File: prow/plugins/lgtm/lgtm.go [hunk 7]
```diff
@@ -204,3 +207,34 @@ func handle(gc githubClient, log *logrus.Entry, e *event) error {
...
+type ghLabelClient interface {
+	RemoveLabel(owner, repo string, number int, label string) error
+}
+
+func handlePullRequest(gc ghLabelClient, pe github.PullRequestEvent) error {
+	if pe.PullRequest.Merged {
+		return nil
+	}
+
+	if pe.Action != "synchronize" {
```
Comment (spxtr) [inline prow/plugins/lgtm/lgtm.go]: This is now `github.PullRequestActionSynchronize`.

File: prow/plugins/lgtm/lgtm_test.go [hunk 8]
```diff
@@ -314,3 +314,105 @@ func TestPRLabelChecker(t *testing.T) {
...
+	return c.err
+}
+
+func TestHandlePullRequest(t *testing.T) {
+	cases := []struct {
+		name           string
+		event          github.PullRequestEvent
+		removeLabelErr error
+
+		err           error
```
Comment (spxtr) [inline prow/plugins/lgtm/lgtm_test.go]: I don't think you have a testcase that actually uses this.

File: prow/plugins/lgtm/lgtm_test.go [hunk 9]
```diff
@@ -314,3 +314,105 @@ func TestPRLabelChecker(t *testing.T) {
...
+
+			if err != nil && c.err == nil {
+				t.Fatalf("handlePullRequest error: %v", err)
+			}
+
+			if err == nil && c.err != nil {
+				t.Fatalf("handlePullRequest wanted error: %v, got nil", c.err)
+			}
+
+			if got, want := err, c.err; got != want {
```
Comment (spxtr) [inline prow/plugins/lgtm/lgtm_test.go]: I don't think this would work even if you had a testcase using it, since they might not be equal but they might still be equivalent.

**Comment by nlandolfi** [comment]
update: this is on my radar just a bit crammed at school right now

**Comment by cblecker** [comment]
@nlandolfi Is this still on your radar? If you don't have cycles for it, we may have to carry this PR.

**Comment by BenTheElder** [comment]
@cjwagner we should probably get something like this in if we haven't already

**Comment by nlandolfi** [comment]
@cblecker -- shoot dropped the ball on this, just tried to do a rebase but (not surprisingly) a bit has changed. 

Y'all might be better positioned to lock this in than me, since I don't have it all in cache at the moment. Thoughts?

**Comment by kargakis** [comment]
Obsoleted by https://github.com/kubernetes/test-infra/pull/6886
